{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in data\n",
    "In this tutorial, it required four input data:\n",
    "\n",
    "1. \"HE_WSI.jpg\": High-resolution Whole-slide H&E stained image \n",
    "2. \"filtered_feature_bc_matrix.h5\": The gene expression matrix\n",
    "3. \"tissue_positions_list.csv\": Spatial coordinateds\n",
    "The gene expreesion data can be stored as an AnnData object. AnnData stores a data matrix .X together with annotations of observations .obs and variables .var\n",
    "\n",
    "The files \"HE_WSI.jpg\" will be used for ResNet morphological feature extraction, \"tissue_positions_list.csv\" will be utilized in the image crop process and feature integration, while the file \"filtered_feature_bc_matrix.h5\" will be used in the gene expression feature matrix module."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Resnet feature extraction\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = '../sample/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import packages\n",
    "import os\n",
    "import cv2\n",
    "import re\n",
    "import sys\n",
    "import fileinput\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import skimage.io\n",
    "\n",
    "img_path=os.path.join(path,'image')\n",
    "img_ori = cv2.imread(img_path)\n",
    "\n",
    "if img_ori is None:\n",
    "    print(\"Failed to load image\")\n",
    "else:\n",
    "    print(img_ori.shape)\n",
    "\n",
    "img = cv2.cvtColor(img_ori, cv2.COLOR_BGR2RGB)\n",
    "plt.figure(dpi=80)\n",
    "plt.imshow(img)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The primary stage of image preprocessing involved segmenting whole-slide H&E histological images from patient samples into smaller patches. \n",
    "Patches were selected such that each one entirely contained the corresponding spots under the tissue, as indicated in the second column of the 'tissue_positions_list.csv' if it is 1. \n",
    "For each spot Si, a corresponding patch was defined, with the center of the patch aligned with the center of the spot, denoted by coordinates (xi, yi) from the last two columns in 'tissue_positions_list.csv'. \n",
    "The dimensions of the patch were such that both the height and width were equal to the diameter of the spot, d, as derived from 'scalefactors_json.json'. \n",
    "Thus, the boundaries of each patch were determined by the following coordinates: \n",
    "the upper boundary at (xi+d/2), the lower boundary at (xi-d/2), the left boundary at (yi-d/2), and the right boundary at (yi + d/2)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "r = 100 #### Need to check diameter EVERYTIME \n",
    "for line in open(os.path.join(path, \"spatial\", \"tissue_positions_list.csv\"), 'r'):\n",
    "  if line.startswith('b') or int(line.rstrip().split(',')[1]) == 0:\n",
    "    continue\n",
    "  x = int(line.rstrip().split(',')[5])\n",
    "  y = int(line.rstrip().split(',')[4])\n",
    "  # print(x)\n",
    "  # print(y)\n",
    "  tiles = image_copy[y-r:y+r, x-r:x+r]\n",
    "  cv2.imwrite(os.path.join(path,'image') +str(x)+'-'+str(y)+'.jpg', tiles)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We utilized a pretrained ResNet-50 model (Tensorflow version: 2.6.0), which was initially trained on the ImageNet dataset, to ensure optimal performance in our task. \n",
    "Specifically, we used it with the top fully-connected layer excluded and selected \"avg\" for the pooling mode to perform feature extraction. \n",
    "Apart from these, everything else was set to the default settings.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from tensorflow.keras.applications.resnet50 import ResNet50\n",
    "from tensorflow.keras.models import Sequential\n",
    "\n",
    "resnet = ResNet50(include_top=False, pooling='avg', weights='imagenet', input_shape=(224, 224, 3))\n",
    "my_new_model = Sequential()\n",
    "my_new_model.add(resnet)\n",
    "# Say not to train first layer (ResNet) model. It is already trained\n",
    "my_new_model.layers[0].trainable = False\n",
    "my_new_model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%time\n",
    "from tensorflow.keras.applications.resnet50 import preprocess_input\n",
    "import cv2 \n",
    "import numpy as np\n",
    "\n",
    "resnet_feature_list = []\n",
    "images = [f for f in os.listdir(DATA_DIR)]\n",
    "\n",
    "for image in images:\n",
    "    file = DATA_DIR+image\n",
    "    \n",
    "    im = cv2.imread(file)\n",
    "    im = cv2.resize(im,(224,224))\n",
    "    img = preprocess_input(np.expand_dims(im.copy(), axis=0))\n",
    "    resnet_feature = my_new_model.predict(img)\n",
    "    resnet_feature_np = np.array(resnet_feature)\n",
    "    resnet_feature_list.append(resnet_feature_np.flatten())\n",
    "\n",
    "array = np.array(resnet_feature_list)\n",
    "np.save(os.path.join(path,'image', 'resnet_array.npy'), array) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "scaler = StandardScaler()\n",
    "array_normalized = scaler.fit_transform(array)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gene expression feature exraction"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Raw gene expression data underwent Counts Per Million (CPM) normalization, and subsequent log transformation followed by scaling the data to have zero mean and unit variance. All these steps were completed using Scanpy (version:1.6.0). \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Read original 10x_h5 data and save it to h5ad\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "from scanpy import read_10x_h5\n",
    "adata = read_10x_h5(os.path.join(path, \"filtered_feature_bc_matrix.h5\"))\n",
    "adata.var_names_make_unique()\n",
    "spatial=pd.read_csv(os.path.join(path, \"spatial\", \"tissue_positions_list.csv\"), sep=\",\", header=None, na_filter=False,index_col=0) \n",
    "adata.obs[\"x1\"]=spatial[1]\n",
    "\n",
    "#Select captured samples\n",
    "adata=adata[adata.obs[\"x1\"]==1]\n",
    "adata.var_names=[i.upper() for i in list(adata.var_names)]\n",
    "adata.var[\"genename\"]=adata.var.index.astype(\"str\")\n",
    "adata.write_h5ad(os.path.join(path, \"sample.h5ad\"))\n",
    "\n",
    "#Read in gene expression and spatial location\n",
    "adata=sc.read(os.path.join(path, \"sample.h5ad\"))\n",
    "\n",
    "# Perform CPM normalization\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "# Log-transform the data\n",
    "sc.pp.log1p(adata)\n",
    "sc.pp.scale(adata)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Resnet morphological and gene expression feature integration"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Both the normalized gene expression matrix and morphological feature matrix underwent Principal Component Analysis (PCA), separately processed through the top 10 principal components. The resulting matrices were concatenated based on the corresponding barcode of the spots."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "DATA_DIR = os.path.join(path, \"image\")\n",
    "images = [f for f in os.listdir(DATA_DIR)]\n",
    "img_ids_list = [f[0:-4] for f in images]\n",
    "\n",
    "n = np.zeros((len(img_ids_list), 2))\n",
    "i = 0\n",
    "j = 0\n",
    "\n",
    "for vector in img_ids_list:\n",
    "  x = int(vector.split('-')[0])\n",
    "  y = int(vector.split('-')[1])\n",
    "  n[i][0]=y\n",
    "  n[i][1]=x\n",
    "  i = i + 1\n",
    "\n",
    "image_ids = pd.DataFrame(n)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Resnet50 PCA\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "pca = PCA(n_components=10)\n",
    "resne_fea = pca.fit_transform(array)\n",
    "resne_fea = pd.DataFrame(resne_fea)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# resnet features with ids\n",
    "resnet_feature_matrix_combined = pd.concat([image_ids, resne_fea], axis=1, ignore_index=True)\n",
    "resnet_feature_matrix_combined.rename(columns={0:'imagerow', 1:'imagecol'}, inplace=True )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>imagerow</th>\n",
       "      <th>imagecol</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TAAGTAAATGTGCCGC-1</td>\n",
       "      <td>1610</td>\n",
       "      <td>6713</td>\n",
       "      <td>-0.625754</td>\n",
       "      <td>2.208479</td>\n",
       "      <td>-14.621526</td>\n",
       "      <td>-3.867671</td>\n",
       "      <td>-4.253917</td>\n",
       "      <td>2.006824</td>\n",
       "      <td>-1.681317</td>\n",
       "      <td>-7.761388</td>\n",
       "      <td>2.961636</td>\n",
       "      <td>0.106228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TGGCACGAGCTCGAGT-1</td>\n",
       "      <td>1609</td>\n",
       "      <td>6831</td>\n",
       "      <td>-1.132644</td>\n",
       "      <td>11.041162</td>\n",
       "      <td>-5.307541</td>\n",
       "      <td>-4.020304</td>\n",
       "      <td>0.948846</td>\n",
       "      <td>-4.475143</td>\n",
       "      <td>-1.311630</td>\n",
       "      <td>-3.712501</td>\n",
       "      <td>6.510522</td>\n",
       "      <td>0.453041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACCGGTCTGAGTACGG-1</td>\n",
       "      <td>1507</td>\n",
       "      <td>6889</td>\n",
       "      <td>-15.165995</td>\n",
       "      <td>-6.775286</td>\n",
       "      <td>-7.886725</td>\n",
       "      <td>-5.320229</td>\n",
       "      <td>-4.165866</td>\n",
       "      <td>3.182706</td>\n",
       "      <td>3.910902</td>\n",
       "      <td>2.281719</td>\n",
       "      <td>2.217093</td>\n",
       "      <td>-1.745553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GAACTTAGCGCCCGGT-1</td>\n",
       "      <td>1609</td>\n",
       "      <td>6948</td>\n",
       "      <td>-6.986549</td>\n",
       "      <td>4.449296</td>\n",
       "      <td>-6.655492</td>\n",
       "      <td>-0.533981</td>\n",
       "      <td>-3.192317</td>\n",
       "      <td>-5.359953</td>\n",
       "      <td>0.268802</td>\n",
       "      <td>-4.441774</td>\n",
       "      <td>7.281716</td>\n",
       "      <td>-2.648497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGTAGCTAGACGCCGA-1</td>\n",
       "      <td>1507</td>\n",
       "      <td>7007</td>\n",
       "      <td>-11.896353</td>\n",
       "      <td>-4.481702</td>\n",
       "      <td>-6.624698</td>\n",
       "      <td>-2.810797</td>\n",
       "      <td>-2.488305</td>\n",
       "      <td>-2.267740</td>\n",
       "      <td>5.924125</td>\n",
       "      <td>4.250512</td>\n",
       "      <td>3.352345</td>\n",
       "      <td>-6.920331</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              barcode  imagerow  imagecol          2          3          4  \\\n",
       "0  TAAGTAAATGTGCCGC-1      1610      6713  -0.625754   2.208479 -14.621526   \n",
       "1  TGGCACGAGCTCGAGT-1      1609      6831  -1.132644  11.041162  -5.307541   \n",
       "2  ACCGGTCTGAGTACGG-1      1507      6889 -15.165995  -6.775286  -7.886725   \n",
       "3  GAACTTAGCGCCCGGT-1      1609      6948  -6.986549   4.449296  -6.655492   \n",
       "4  AGTAGCTAGACGCCGA-1      1507      7007 -11.896353  -4.481702  -6.624698   \n",
       "\n",
       "          5         6         7         8         9        10        11  \n",
       "0 -3.867671 -4.253917  2.006824 -1.681317 -7.761388  2.961636  0.106228  \n",
       "1 -4.020304  0.948846 -4.475143 -1.311630 -3.712501  6.510522  0.453041  \n",
       "2 -5.320229 -4.165866  3.182706  3.910902  2.281719  2.217093 -1.745553  \n",
       "3 -0.533981 -3.192317 -5.359953  0.268802 -4.441774  7.281716 -2.648497  \n",
       "4 -2.810797 -2.488305 -2.267740  5.924125  4.250512  3.352345 -6.920331  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tissue_position = pd.read_csv((os.path.join(path, \"spatial\", \"tissue_positions_list.csv\")), sep=',', names=[\"barcode\", \"tissue\", \"row\", \"col\",\"imagerow\", \"imagecol\"])\n",
    "combined_feature_matirx = pd.merge(tissue_position, resnet_feature_matrix_combined, how = \"inner\", on = ['imagerow', 'imagecol'])\n",
    "resnet_remix = combined_feature_matirx.drop([\"tissue\", \"row\", \"col\"], axis=1)\n",
    "\n",
    "resnet_remix.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Gene expression in Adata PCA\n",
    "gene_exp = adata.to_df()\n",
    "pca = PCA(n_components=10)\n",
    "embed = pca.fit_transform(gene_exp)\n",
    "\n",
    "exp_emb = pd.DataFrame(embed)\n",
    "exp_emb['barcode'] = df.index\n",
    "exp_emb['barcode'] = exp_emb['barcode'].str.decode('utf8')\n",
    "barcode = exp_emb[\"barcode\"]\n",
    "exp_emb = exp_emb.drop(columns=['barcode'])\n",
    "exp_emb.insert(loc=0, column='barcode', value=barcode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>2_x</th>\n",
       "      <th>3_x</th>\n",
       "      <th>4_x</th>\n",
       "      <th>5_x</th>\n",
       "      <th>6_x</th>\n",
       "      <th>7_x</th>\n",
       "      <th>8_x</th>\n",
       "      <th>9_x</th>\n",
       "      <th>10</th>\n",
       "      <th>...</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2_y</th>\n",
       "      <th>3_y</th>\n",
       "      <th>4_y</th>\n",
       "      <th>5_y</th>\n",
       "      <th>6_y</th>\n",
       "      <th>7_y</th>\n",
       "      <th>8_y</th>\n",
       "      <th>9_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TAAGTAAATGTGCCGC-1</td>\n",
       "      <td>-0.625754</td>\n",
       "      <td>2.208479</td>\n",
       "      <td>-14.621526</td>\n",
       "      <td>-3.867671</td>\n",
       "      <td>-4.253917</td>\n",
       "      <td>2.006824</td>\n",
       "      <td>-1.681317</td>\n",
       "      <td>-7.761388</td>\n",
       "      <td>2.961636</td>\n",
       "      <td>...</td>\n",
       "      <td>29.329668</td>\n",
       "      <td>10.181683</td>\n",
       "      <td>-4.020726</td>\n",
       "      <td>7.109770</td>\n",
       "      <td>1.738399</td>\n",
       "      <td>-5.661458</td>\n",
       "      <td>-0.715062</td>\n",
       "      <td>1.739450</td>\n",
       "      <td>3.392922</td>\n",
       "      <td>-2.633126</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TGGCACGAGCTCGAGT-1</td>\n",
       "      <td>-1.132644</td>\n",
       "      <td>11.041162</td>\n",
       "      <td>-5.307541</td>\n",
       "      <td>-4.020304</td>\n",
       "      <td>0.948846</td>\n",
       "      <td>-4.475143</td>\n",
       "      <td>-1.311630</td>\n",
       "      <td>-3.712501</td>\n",
       "      <td>6.510522</td>\n",
       "      <td>...</td>\n",
       "      <td>29.114092</td>\n",
       "      <td>4.093781</td>\n",
       "      <td>3.140975</td>\n",
       "      <td>0.607610</td>\n",
       "      <td>-2.490068</td>\n",
       "      <td>-1.254769</td>\n",
       "      <td>-0.112335</td>\n",
       "      <td>-2.849063</td>\n",
       "      <td>-2.263700</td>\n",
       "      <td>-5.127917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ACCGGTCTGAGTACGG-1</td>\n",
       "      <td>-15.165995</td>\n",
       "      <td>-6.775286</td>\n",
       "      <td>-7.886725</td>\n",
       "      <td>-5.320229</td>\n",
       "      <td>-4.165866</td>\n",
       "      <td>3.182706</td>\n",
       "      <td>3.910902</td>\n",
       "      <td>2.281719</td>\n",
       "      <td>2.217093</td>\n",
       "      <td>...</td>\n",
       "      <td>32.618317</td>\n",
       "      <td>1.488421</td>\n",
       "      <td>3.387632</td>\n",
       "      <td>-1.013952</td>\n",
       "      <td>-0.396817</td>\n",
       "      <td>-5.833042</td>\n",
       "      <td>-2.133456</td>\n",
       "      <td>0.479996</td>\n",
       "      <td>2.592476</td>\n",
       "      <td>-3.428612</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>GAACTTAGCGCCCGGT-1</td>\n",
       "      <td>-6.986549</td>\n",
       "      <td>4.449296</td>\n",
       "      <td>-6.655492</td>\n",
       "      <td>-0.533981</td>\n",
       "      <td>-3.192317</td>\n",
       "      <td>-5.359953</td>\n",
       "      <td>0.268802</td>\n",
       "      <td>-4.441774</td>\n",
       "      <td>7.281716</td>\n",
       "      <td>...</td>\n",
       "      <td>10.097507</td>\n",
       "      <td>0.669171</td>\n",
       "      <td>-5.001672</td>\n",
       "      <td>-5.451578</td>\n",
       "      <td>-7.546212</td>\n",
       "      <td>4.614475</td>\n",
       "      <td>0.163925</td>\n",
       "      <td>-4.826335</td>\n",
       "      <td>0.888212</td>\n",
       "      <td>-2.120166</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AGTAGCTAGACGCCGA-1</td>\n",
       "      <td>-11.896353</td>\n",
       "      <td>-4.481702</td>\n",
       "      <td>-6.624698</td>\n",
       "      <td>-2.810797</td>\n",
       "      <td>-2.488305</td>\n",
       "      <td>-2.267740</td>\n",
       "      <td>5.924125</td>\n",
       "      <td>4.250512</td>\n",
       "      <td>3.352345</td>\n",
       "      <td>...</td>\n",
       "      <td>27.485001</td>\n",
       "      <td>4.553941</td>\n",
       "      <td>5.455178</td>\n",
       "      <td>-4.433626</td>\n",
       "      <td>-5.616123</td>\n",
       "      <td>-1.932594</td>\n",
       "      <td>2.287708</td>\n",
       "      <td>-0.626706</td>\n",
       "      <td>0.964270</td>\n",
       "      <td>-3.938417</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              barcode        2_x        3_x        4_x       5_x       6_x  \\\n",
       "0  TAAGTAAATGTGCCGC-1  -0.625754   2.208479 -14.621526 -3.867671 -4.253917   \n",
       "1  TGGCACGAGCTCGAGT-1  -1.132644  11.041162  -5.307541 -4.020304  0.948846   \n",
       "2  ACCGGTCTGAGTACGG-1 -15.165995  -6.775286  -7.886725 -5.320229 -4.165866   \n",
       "3  GAACTTAGCGCCCGGT-1  -6.986549   4.449296  -6.655492 -0.533981 -3.192317   \n",
       "4  AGTAGCTAGACGCCGA-1 -11.896353  -4.481702  -6.624698 -2.810797 -2.488305   \n",
       "\n",
       "        7_x       8_x       9_x        10  ...          0          1  \\\n",
       "0  2.006824 -1.681317 -7.761388  2.961636  ...  29.329668  10.181683   \n",
       "1 -4.475143 -1.311630 -3.712501  6.510522  ...  29.114092   4.093781   \n",
       "2  3.182706  3.910902  2.281719  2.217093  ...  32.618317   1.488421   \n",
       "3 -5.359953  0.268802 -4.441774  7.281716  ...  10.097507   0.669171   \n",
       "4 -2.267740  5.924125  4.250512  3.352345  ...  27.485001   4.553941   \n",
       "\n",
       "        2_y       3_y       4_y       5_y       6_y       7_y       8_y  \\\n",
       "0 -4.020726  7.109770  1.738399 -5.661458 -0.715062  1.739450  3.392922   \n",
       "1  3.140975  0.607610 -2.490068 -1.254769 -0.112335 -2.849063 -2.263700   \n",
       "2  3.387632 -1.013952 -0.396817 -5.833042 -2.133456  0.479996  2.592476   \n",
       "3 -5.001672 -5.451578 -7.546212  4.614475  0.163925 -4.826335  0.888212   \n",
       "4  5.455178 -4.433626 -5.616123 -1.932594  2.287708 -0.626706  0.964270   \n",
       "\n",
       "        9_y  \n",
       "0 -2.633126  \n",
       "1 -5.127917  \n",
       "2 -3.428612  \n",
       "3 -2.120166  \n",
       "4 -3.938417  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resnet_plus_exp_PCs = pd.merge(resnet_remix, exp_emb, how = \"inner\", on = ['barcode']).drop(columns=['imagerow', 'imagecol'])\n",
    "resnet_plus_exp_PCs.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Note\n",
    "Clustering was accomplished using the Louvain modularity optimization algorithm with a resolution range of 1.5-1.9 and k-neighbors set at 39, implemented in Orange Data Mining (version 3.30.1). Make adjustment on resolution and k neighbors to make sure the number of clusters in output equals to the number of clusters from default Louvain clustering on just gene expression (you can find this from either loupe browser or spaceranger /out/analysis/clustering/graphclust/clusters.csv)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "for_clustering_input = resnet_plus_exp_PCs.iloc[:,1:]\n",
    "# save the output from pervious steps to Orange Data Mining\n",
    "np.savetxt((os.path.join(path, \"annotation\", \"for_clustering_input.csv\")), for_clustering_input, delimiter=',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in the output from \n",
    "orange_output = pd.read_csv(os.path.join(path, \"annotation\", \"resnet_plus_exp_clustering_v2.csv\"), sep=',', skiprows=3, usecols = [20], header = None)\n",
    "cluster_label = pd.concat([resnet_plus_exp_PCs, orange_output], axis=1, ignore_index=True)\n",
    "cluster_label = cluster_label.iloc[:, [0, -1]]\n",
    "cluster_label.rename( columns={cluster_label.columns[0] :'barcode', cluster_label.columns[-1]:'cluster'}, inplace=True )\n",
    "cluster_label['cluster'] = cluster_label['cluster'].str.replace(r'\\D', '').astype(int)\n",
    "cluster_label.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>barcode</th>\n",
       "      <th>tissue</th>\n",
       "      <th>row</th>\n",
       "      <th>col</th>\n",
       "      <th>imagerow</th>\n",
       "      <th>imagecol</th>\n",
       "      <th>cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ACGCCTGACACGCGCT-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1111</td>\n",
       "      <td>1000</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TACCGATCCAACACTT-1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1213</td>\n",
       "      <td>1059</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ATTAAAGCGGACGAGC-1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1111</td>\n",
       "      <td>1118</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              barcode  tissue  row  col  imagerow  imagecol  cluster\n",
       "0  ACGCCTGACACGCGCT-1       0    0    0      1111      1000      NaN\n",
       "1  TACCGATCCAACACTT-1       0    1    1      1213      1059      NaN\n",
       "2  ATTAAAGCGGACGAGC-1       0    0    2      1111      1118      NaN"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merge = tissue_position.merge(cluster_label, how = 'outer', on = ['barcode'])\n",
    "louvain_cluster = merge[['barcode', 'cluster']]\n",
    "merge = merge.astype(float)\n",
    "louvain_cluster = louvain_cluster.astype(float)\n",
    "np.savetxt((os.path.join(path, \"annotation\", \"R_E_louvain.csv\")), merge, delimiter=',')\n",
    "np.savetxt((os.path.join(path, \"annotation\", \"R_E_cluster.csv\")), louvain_cluster, delimiter=',')\n",
    "merge.head(3)"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "78fc04bcb9a9464107d2091a715d17e26e41bd40e804d5ee3bc14168bce6b7a5"
  },
  "kernelspec": {
   "display_name": "Python 3.7.6 64-bit ('base': conda)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
